src/lambda/expression.rs

```diff
@@ -181,11 +206,76 @@ impl Expression {
             result
           }
         },
+        Expression::Cache(cache) => {
+          if let Some(key) = get_cache_key(ctx, DefaultHasher::new()) {
```

Comment:
I don't think here could be simple use of `DefaultHasher`as we lose a lot of information on parent types. Take a look into `src/blueprint/into_schema.rs` file and how is hasher is used there. We need to separate fields with the same name from different types I believe.

Anyway, we need a help from @shashitnak to verify this for cache

Reply:
yes `DefaultHasher::new()` shouldn't be be used here. This hasher needs to be cloned from `cache` field in `FieldDefinition`. It has a `DefaultHasher` which has already hashed the name of the def and the field

Reply:
@shashitnak Could you also write the steps that are important for hashing here for reference.
Is it smth like `TypeName -> FieldName -> get_cache_key`?

Reply:
yes, that is correct. And `get_cache_key` then hashes on the value of `id` field if present in the parent of field and on the name and value of arguments of the query

Reply:
@shashitnak tests were passed for this. Could you check the latest changes of the pr to verify it's ok? or maybe you can suggest how to test it properly and what kind of test should be added to catch this?

Reply:
This is a test which can catch this bug. I have tested it with the caching implementation in the main branch and it fails when I provide `DefaultHasher::new()` to `get_cache_key` and passes when I pass the hasher from the `Cache` object.

`.graphql` file
```
schema @upstream(baseURL: "http://example.com", batch: {delay: 1, maxSize: 1000}) {
  query: Query
}

type Query {
  foo: Foo @http(path: "/foo")
}

type Foo {
  id: Int!
  bar1: Bar @http(path: "/bar1") @cache(maxAge: 1000)
  bar2: Bar @http(path: "/bar2") @cache(maxAge: 1000)
}

type Bar {
  id: Int!
}
```

`.yml` file
```
config: !file tests/http/config/caching-gql-collision.graphql
name: Caching Graphql args
mock:
  - request:
      url: http://example.com/foo
    response:
      body:
        id: 0
  - request:
      url: http://example.com/bar1
    response:
      body:
        id: 1
  - request:
      url: http://example.com/bar2
    response:
      body:
        id: 2
assert:
  - request:
      method: POST
      url: http://localhost:8080/graphql
      body:
        query: "query { foo { id bar1 { id } bar2 { id } } }"
    response:
      body:
        data:
          foo:
            id: 0
            bar1:
              id: 1
            bar2:
              id: 2
```
The idea is since name of the type and the field is not used, `bar1` and `bar2` result in the same hash and the result of `bar1` is returned for `bar2` from the cache, hence the test fails.

@Shylock-Hg Please add this in the tests. If you find any mistake in the test or if it is not catching this bug correctly let me know

Reply:
I've extended the test file with more check, please, take a look https://github.com/tailcallhq/tailcall/pull/833/files#diff-4ae3c7a1b1832e15a94d554d6610f6b51427fff9c090ef7ab77b7c8f2172406b

Reply:
I didn't get the question, could you rephrase it?

Reply:
ok, got it. 

Since that test file didn't catch the problem during working on that pr I have doubts that it is useful at all. Also, that was a big file that was hard to read and understand what is going on so it decrease its value. New test catches several cases and it is useful already at least